src/golang/cmd/migrator/versions/000016_add_artifact_type_column_to_artifact/utils.go

```diff
@@ -0,0 +1,315 @@
+	}
+
+	return (&n.Metadata).Value()
+}
+
+func (n *NullMetadata) Scan(value interface{}) error {
+	if value == nil {
+		n.IsNull = true
+		return nil
+	}
+
+	metadata := &Metadata{}
+	if err := metadata.Scan(value); err != nil {
+		return err
+	}
+
+	n.Metadata, n.IsNull = *metadata, false
+	return nil
+}
+
+type artifactResult struct {
+	Id          uuid.UUID    `db:"id"`
+	Metadata    NullMetadata `db:"metadata"`
+	ContentPath string       `db:"content_path" json:"content_path"`
+}
+
+func getArtifactResult(ctx context.Context, db database.Database, artifactId uuid.UUID) ([]artifactResult, error) {
+	query := "SELECT id, metadata, content_path FROM artifact_result WHERE artifact_id = $1;"
+
+	var result []artifactResult
+	err := db.Query(ctx, &result, query, artifactId)
+	return result, err
+}
+
+func updateMetadataInArtifactResult(
+	ctx context.Context,
+	id uuid.UUID,
+	metadata *NullMetadata,
+	db database.Database,
+) error {
+	changes := map[string]interface{}{
+		"metadata": metadata,
+	}
+	return utils.UpdateRecord(ctx, changes, "artifact_result", "id", id, db)
+}
+
+type MigrationSpec struct {
+	ArtifactType  string               `json:"artifact_type"`
+	StorageConfig shared.StorageConfig `json:"storage_config"`
+	MetadataPath  string               `json:"metadata_path"`
+	ContentPath   string               `json:"content_path"`
+}
+
+type NewArtifactType string
+
+type SerializationType string
+
+type TypeMetadata struct {
+	ArtifactType      NewArtifactType   `json:"artifact_type"`
+	SerializationType SerializationType `json:"serialization_type"`
+}
+
+func updateTypeInArtifact(
+	ctx context.Context,
+	id uuid.UUID,
+	artifactType NewArtifactType,
+	db database.Database,
+) error {
+	changes := map[string]interface{}{
+		"type": artifactType,
+	}
+	return utils.UpdateRecord(ctx, changes, "artifact", "id", id, db)
+}
+
+func migrateArtifact(ctx context.Context, db database.Database) error {
+	artifactResultMigrated := 0
+
+	artifactSpecs, err := getArtifactSpec(ctx, db)
+	if err != nil {
+		return err
+	}
+
+	for _, artifactSpec := range artifactSpecs {
+		artifactResults, err := getArtifactResult(ctx, db, artifactSpec.Id)
+		if err != nil {
+			return err
+		}
+
+		newArtifactType := NewArtifactType("")
+
+		for _, artifactResult := range artifactResults {
+			// Temporaty file to store the updated metadata dict that contains
+			// the serialization type and artifact type.
+			metadataPath := fmt.Sprintf("%s_%s", artifactResult.Id, "metadata")
+			storageConfig := config.ParseServerConfiguration(confPath).StorageConfig
+
+			migrationSpec := MigrationSpec{
+				ArtifactType:  string(artifactSpec.Spec.spec.Type),
+				StorageConfig: *storageConfig,
+				MetadataPath:  metadataPath,
+				ContentPath:   artifactResult.ContentPath,
+			}
+
+			specData, err := json.Marshal(migrationSpec)
+			if err != nil {
+				return err
+			}
+
+			// Launch the Python migration job with the spec constructed above.
+			cmd := exec.Command(
```

Comment:
Do we have any sort of a storage rollback mechanism if the storage migration is successful, but the db migration fails?

Reply:
Yeah good catch. Let me copy the original content file to a temp file, and if the database migration fails we can replace the storage with the original version.

Reply:
added